native/com_wolfssl_WolfSSL.c

```diff
@@ -678,6 +686,138 @@ void NativeLoggingCallback(const int logLevel, const char *const logMessage)
     }
 }
 
+void NativeFIPSErrorCallback(const int ok, const int err,
+                             const char* const hash)
+{
+#ifdef HAVE_FIPS
+    JNIEnv*   jenv;
+    jint      vmret  = 0;
+    jclass    excClass;
+    jmethodID errorMethod;
+    jobjectRefType refcheck;
+
+    /* get JNIEnv from JavaVM */
+    vmret = (int)((*g_vm)->GetEnv(g_vm, (void**) &jenv, JNI_VERSION_1_6));
+    if (vmret == JNI_EDETACHED) {
+#ifdef __ANDROID__
+        vmret = (*g_vm)->AttachCurrentThread(g_vm, &jenv, NULL);
+#else
+        vmret = (*g_vm)->AttachCurrentThread(g_vm, (void**) &jenv, NULL);
+#endif
+        if (vmret) {
+            printf("Failed to attach JNIEnv to thread\n");
+        }
+    } else if (vmret != JNI_OK) {
+        printf("Unable to get JNIEnv from JavaVM\n");
+    }
+
+    /* find exception class */
+    excClass = (*jenv)->FindClass(jenv, "java/lang/Exception");
+    if ((*jenv)->ExceptionOccurred(jenv)) {
+        (*jenv)->ExceptionDescribe(jenv);
+        (*jenv)->ExceptionClear(jenv);
+        return;
+    }
+
+    /* check if our stored object reference is valid */
+    refcheck = (*jenv)->GetObjectRefType(jenv, g_fipsCbIfaceObj);
+    if (refcheck == 2) {
+
+        /* lookup WolfSSLLoggingCallback class from global object ref */
+        jclass fipsCbClass = (*jenv)->GetObjectClass(jenv, g_fipsCbIfaceObj);
+        if (!fipsCbClass) {
+            if ((*jenv)->ExceptionOccurred(jenv)) {
+                (*jenv)->ExceptionDescribe(jenv);
+                (*jenv)->ExceptionClear(jenv);
+            }
+
+            (*jenv)->ThrowNew(jenv, excClass,
+                "Can't get native WolfSSLFIPSErrorCallback class reference");
+            return;
+        }
+
+        errorMethod = (*jenv)->GetMethodID(jenv, fipsCbClass,
+                                            "errorCallback",
+                                            "(IILjava/lang/String;)V");
+        if (errorMethod == 0) {
+            if ((*jenv)->ExceptionOccurred(jenv)) {
+                (*jenv)->ExceptionDescribe(jenv);
+                (*jenv)->ExceptionClear(jenv);
+            }
+            (*jenv)->ThrowNew(jenv, excClass,
+                "Error getting errorCallback method from JNI");
+            return;
+        }
+
+        /* create jstring from char* */
+        jstring hashString = (*jenv)->NewStringUTF(jenv, hash);
```

Comment:
does this jstring need free'd?

Reply:
Normally if we were returning to Java from this function, the local references would automatically be cleaned up.  Since this is returning back to native, you are correct.  Added a call to DeleteLocalRef() for hashString.